sqlmodel/main.py

```diff
@@ -562,7 +562,31 @@ def get_config(name: str) -> Any:
             # If it was passed by kwargs, ensure it's also set in config
             set_config_value(model=new_cls, parameter="table", value=config_table)
             for k, v in get_model_fields(new_cls).items():
-                col = get_column_from_field(v)
+                if not IS_PYDANTIC_V2:
+                    col = get_column_from_field(v)
+                    setattr(new_cls, k, col)
+                    continue
+
+                original_field = getattr(v, "_original_assignment", Undefined)
+                # Get the original sqlmodel FieldInfo, pydantic >=v2.12 changes the model
+                if isinstance(original_field, FieldInfo):
+                    field = original_field
+                else:
+                    annotated_field = next(
+                        ann
+                        for c in new_cls.__mro__
+                        if (ann := get_annotations(c.__dict__).get(k))  # type: ignore[arg-type]
+                    )
+                    annotated_field_meta = getattr(
+                        annotated_field, "__metadata__", (())
+                    )
+                    field = next(
+                        (f for f in annotated_field_meta if isinstance(f, FieldInfo)),
+                        v,
+                    )  # type: ignore[assignment]
+                field.annotation = v.annotation
+                # Guarantee the field has the correct type
+                col = get_column_from_field(field)
```

Comment:
```suggestion
                else:
                    original_field = getattr(v, "_original_assignment", Undefined)
                    if isinstance(original_field, FieldInfo):
                        field = original_field
                    else:
                        field = find_field_info(new_cls, k, v)
                    field.annotation = v.annotation
                    col = get_column_from_field(field)
                    setattr(new_cls, k, col)
```

I think this part will be clearer with `if..else` instead of `continue` and with some logic moved to a separate function:
```py
def find_field_info(cls: type, field_name: str, fallback: FieldInfo) -> FieldInfo:
    for c in cls.__mro__:
        annotated = get_annotations(c.__dict__).get(field_name)  # type: ignore[arg-type]
        if annotated:
            for meta in getattr(annotated, "__metadata__", ()):
                if isinstance(meta, FieldInfo):
                    return meta
    return fallback
```

Reply:
I've further refactored the code according to Yurii's suggestion here.